src/NServiceBus.Core/Transports/Learning/LearningTransportDispatcher.cs

```diff
@@ -0,0 +1,170 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using DelayedDelivery;
+    using DeliveryConstraints;
+    using Extensibility;
+    using Transport;
+
+    class LearningTransportDispatcher : IDispatchMessages
+    {
+        public LearningTransportDispatcher(string basePath)
+        {
+            this.basePath = basePath;
+        }
+
+        public Task Dispatch(TransportOperations outgoingMessages, TransportTransaction transaction, ContextBag context)
+        {
+            return Task.WhenAll(
+                DispatchUnicast(outgoingMessages.UnicastTransportOperations, transaction),
+                DispatchMulticast(outgoingMessages.MulticastTransportOperations, transaction));
+        }
+
+        async Task DispatchMulticast(IEnumerable<MulticastTransportOperation> transportOperations, TransportTransaction transaction)
+        {
+            var tasks = new List<Task>();
+            foreach (var transportOperation in transportOperations)
+            {
+                var subscribers = await GetSubscribersFor(transportOperation.MessageType)
+                    .ConfigureAwait(false);
+
+                foreach (var subscriber in subscribers)
+                {
+                    tasks.Add(WriteMessage(subscriber, transportOperation, transaction));
+                }
+            }
+            await Task.WhenAll(tasks)
+                .ConfigureAwait(false);
+        }
+
+
+        Task DispatchUnicast(IEnumerable<UnicastTransportOperation> operations, TransportTransaction transaction)
+        {
+            return Task.WhenAll(operations.Select(operation => WriteMessage(operation.Destination, operation, transaction)));
```

Comment:
Normally I would advise to avoid LINQ, here it probably doesn't matter. That being said the above method uses foreach instead of Select. Should we use a foreach here as well?